dig/graph.go

```diff
@@ -183,5 +183,38 @@ func (g *graph) registerConstructor(c interface{}) error {
 	}
 
 	g.nodes[objType] = &n
+
+	// object needs to be part of the graph to properly detect cycles
+	if cycleErr := g.detectCycles(&n); cycleErr != nil {
+		// if the cycle was detected delete from the graph
+		delete(g.nodes, objType)
```

Comment:
Yea I see your point as well. I like the approach of keeping Register on initialization. Worth further discussion.